File: PA_16.py
#LC 560 - Subarray Sum Equals K
#Time Complexity - O(n)
#Space Complexity - O(n)
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        d = {0:1}
        ans = 0
        rsum = 0
        for i in range(len(nums)):
            rsum = rsum + nums[i]
            comp = rsum - k
            if comp in d.keys():
                ans = ans + d[comp]
            if rsum not in d.keys():
                d[rsum] = 1
            else:
                d[rsum] = d[rsum] + 1
        return ans

File: PA16.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> hm = new HashMap<>();
        int sum=0;
        int count=0;
        hm.put(0,1);
        for(int i=0;i<nums.length;i++){
            sum= sum+nums[i];
            if(hm.containsKey(sum-k) ){
                count=count+hm.get(sum-k);
            }
            hm.put(sum, hm.getOrDefault(sum,0)+1);
        }
            
        return count;
    }
}